src/line_launcher/mod.rs

```diff
@@ -16,17 +19,17 @@ const VELOCITY: u8 = 100;
 pub struct LineLauncher {
     beat_message_receiver: Receiver<BeatNumber>,
     state: State,
-    notes: Vec<u8>,
-    output: MidiOutputConnection,
+    lines: Vec<Line>,
+    midi_message_sender: MidiMessageSender,
```

Comment:
Inside `listen()`, I was getting borrowing errors since I wanted an immutable borrow of the line from `self.lines` while needing to mutably borrow `self` for eg `self.fire_note_on()`. Tracing it, the only reason `self.fire_note_on()` needed `self` to be mutable was because `self.output` needed to be mutable when actually sending the message

So what I came up with was to refactor to "isolate the mutability" inside a single `midi_message_sender` member (so that `self` doesn't need to be borrowed mutably, just `self.midi_message_sender`)

Reply:
Right, the "mut self makes things tricky" problem comes up quite a bit, especially with long-running functions like this. I think your solution is nice, and also nicely encapsulates some things in the message sender. My general approach to things like this is avoid requiring `&mut self` if at all possible. For a long-running function like this where you're not actually going to use the value of `LineLauncher` after `listen` finishes, you don't really need to mutate `self` (because the value of `self` after the function finishes doesn't matter), so instead I'd pass in anything that needs to be mutated into the `listen` function, rather than including them as fields on the struct:

```diff
From e66c82aee1aafc4fe303453028e4f49ad721ab10 Mon Sep 17 00:00:00 2001
From: Peter Stuart <peter@peterstuart.org>
Date: Sun, 20 Jun 2021 07:41:29 -0400
Subject: [PATCH] Don't require mut self.

---
 src/line_launcher/mod.rs | 62 +++++++++++++++++++++++-----------------
 src/main.rs              |  4 +--
 2 files changed, 37 insertions(+), 29 deletions(-)

diff --git a/src/line_launcher/mod.rs b/src/line_launcher/mod.rs
index 5432ad2..256b527 100644
--- a/src/line_launcher/mod.rs
+++ b/src/line_launcher/mod.rs
@@ -4,7 +4,7 @@ use wmidi::{Channel, MidiMessage, Note, Velocity};
 
 use crate::{lines, BeatNumber, Line};
 
-#[derive(Debug)]
+#[derive(Clone, Copy, Debug)]
 enum State {
     NotPlaying,
     Playing {
@@ -17,43 +17,49 @@ const CHANNEL: Channel = Channel::Ch1;
 const VELOCITY: u8 = 100;
 
 pub struct LineLauncher {
-    beat_message_receiver: Receiver<BeatNumber>,
-    state: State,
     lines: Vec<Line>,
-    midi_message_sender: MidiMessageSender,
 }
 
 impl LineLauncher {
-    pub fn new(beat_message_receiver: Receiver<BeatNumber>, output: MidiOutputConnection) -> Self {
+    pub fn new() -> Self {
         Self {
-            beat_message_receiver,
-            state: State::NotPlaying,
             lines: lines::all_lines(),
-            midi_message_sender: MidiMessageSender { output },
         }
     }
 
-    pub fn listen(&mut self) {
+    pub fn listen(
+        &self,
+        beat_message_receiver: Receiver<BeatNumber>,
+        output: MidiOutputConnection,
+    ) {
+        let mut state = State::NotPlaying;
+        let mut midi_message_sender = MidiMessageSender { output };
+
         loop {
-            let beat_message = self.beat_message_receiver.recv().unwrap();
-            match self.state {
+            let beat_message = beat_message_receiver.recv().unwrap();
+            state = match state {
                 State::NotPlaying if beat_message.is_beginning_of_measure() => {
-                    self.state = State::Playing {
+                    state = State::Playing {
                         line_index: 0,
                         next_note_index: 0,
                     };
-                    self.possibly_trigger_notes(beat_message);
+                    self.possibly_trigger_notes(state, &mut midi_message_sender, beat_message)
                 }
                 State::Playing { .. } => {
-                    self.possibly_trigger_notes(beat_message);
+                    self.possibly_trigger_notes(state, &mut midi_message_sender, beat_message)
                 }
-                _ => {}
+                _ => state,
             }
         }
     }
 
-    fn possibly_trigger_notes(&mut self, beat_message: BeatNumber) {
-        match self.state {
+    fn possibly_trigger_notes(
+        &self,
+        state: State,
+        midi_message_sender: &mut MidiMessageSender,
+        beat_message: BeatNumber,
+    ) -> State {
+        match state {
             State::Playing {
                 line_index,
                 next_note_index,
@@ -65,30 +71,32 @@ impl LineLauncher {
                     if beat_message.minus_sixteenths(last_played_note.duration)
                         == last_played_note.start
                     {
-                        self.midi_message_sender
-                            .fire_note_off(last_played_note.note);
+                        midi_message_sender.fire_note_off(last_played_note.note);
                         did_trigger_note_off = true;
                     }
                 }
                 if next_note_index == line.notes.len() {
-                    if did_trigger_note_off {
-                        self.state = State::NotPlaying;
-                    }
-                    return;
+                    return if did_trigger_note_off {
+                        State::NotPlaying
+                    } else {
+                        state
+                    };
                 }
                 let next_note = &line.notes[next_note_index];
                 if beat_message == next_note.start {
-                    self.midi_message_sender.fire_note_on(next_note.note);
-                    self.state = State::Playing {
+                    midi_message_sender.fire_note_on(next_note.note);
+                    return State::Playing {
                         line_index,
                         next_note_index: next_note_index + 1,
-                    }
+                    };
                 }
+
+                state
             }
             _ => {
                 panic!(
                     "Called possibly_trigger_notes() while not playing: {:?}",
-                    self.state
+                    state
                 );
             }
         }
diff --git a/src/main.rs b/src/main.rs
index 3cb7b4b..71cd548 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -25,8 +25,8 @@ fn main() {
         )
         .unwrap();
 
-    let mut line_launcher = LineLauncher::new(beat_message_receiver, conn_out);
-    line_launcher.listen();
+    let line_launcher = LineLauncher::new();
+    line_launcher.listen(beat_message_receiver, conn_out);
 }
 
 fn handle_message(message: Message, midi_clock_tracker: &mut MidiClockTracker) {
-- 
2.32.0
```

There's a little more boilerplate, as you have to thread some variables through function calls, but it gives you a lot more freedom.

Reply:
Ok that's very handy - updated to mimic your patch. It's nice how there's less side-effect-y `self.state` mutation going on in the middle of everything

Having done that, Clippy encouraged me to use the `Default` trait for `LineLauncher` (I guess because its `new()` wasn't doing anything dynamic), so I did that and removed `new()` at least for now (so used `LineLauncher::default()` in `src/main.rs`) (or maybe what this should really be steering me toward is not needing a `LineLauncher` struct at all?)